draco/synthesis/expand_perturbed.py

```diff
@@ -0,0 +1,598 @@
+                continue
+            elif self.telescope.beamclass[fj] > 1:
+                continue
+
+            # Select visibility corresponding to unique_ind product.
+            prod_stream_noconj = sstream.vis[:, unique_ind]
+            # Conjugate if necessary.
+            prod_stream  = prod_stream_noconj.conj() if conj else prod_stream_noconj
+
+            # Loop over each perturbation, starting from 1 as 0 would be the unperturbed component.
+            for pert_index in range(1,self.telescope.npert):
+                # Define the feed index in sstream for the perturbed component corresponding to each feed.
+                fi_pert = fi + ninputperpert * pert_index
+                fj_pert = fj + ninputperpert * pert_index
+
+                # Define the product index for the fi & fj perturbed component.
+                # By default we won't put this component in new_stream as we're working to first order.
+                # However, its' good to track what it is.
+                pertfifj_ind = self.telescope.feedmap[fi_pert, fj_pert]
+                pertfifj_conj = self.telescope.feedconj[fi_pert,fj_pert]
+
+                # Define product index in sstream for the fi perturbed component.
+                pertfi_ind = self.telescope.feedmap[fi_pert, fj]
+                # Select the visibility corresponding to the fi perturbed component
+                # Multiply it by the perturbation value corresponding to the fi in this perturbation
+                pertfi_stream = sstream.vis[:, pertfi_ind] * feed_perts[fi_pert - ninputperpert * pert_index]
+
+                # Define product index for fj perturbed component
+                pertfj_ind = self.telescope.feedmap[fi, fj_pert]
+                # Select visibility corresponding to fj perturbed component and multiply by perturbation value
+                pertfj_stream = sstream.vis[:, pertfj_ind] * feed_perts[fj - ninputperpert * pert_index]
+
+                # Conjugates if necessary
+                pertfi_conj = self.telescope.feedconj[fi_pert, fj]
+                pertfj_conj = self.telescope.feedconj[fi, fj_pert]
+
+                # Add fi perturbation component to unperturbed sstream
+                prod_stream = prod_stream + (pertfi_stream.conj() if pertfi_conj else pertfi_stream)
+
+                # Add fj perturbation component to unperturbed sstream + fi perturbation sstream
+                prod_stream = prod_stream +  (pertfj_stream.conj() if pertfj_conj else pertfj_stream)
+
+
+            # Put prod_stream into new container new_stream.
+            new_stream.vis[:, pi] = prod_stream
+            new_stream.weight[:, pi] = 1.0
+
+        return new_stream
+
+
+class ExpandPerturbedProducts2ndOrder(task.SingleTask):
+    """Un-wrap collated products to full triangle.
+        Combine perturbed and Unperturbed products, keeping the term perturbed
+        in both feed i and j (which is therefore 2nd order).
+    """
```

Comment:
I echo Carolin's comment for BeamPerturbation and Beam: they share code and maybe could be combined.
I think this is even more true for ExpandPerturbedProducts and ExpandPerturbedProducts2ndOrder. Could you make a single task with an option 'order' to be 1 or 2 set by the yaml file?